### ISSUE TITLE
fpart fails on memory constrained systems

### ISSUE THREAD

**Issue by shenki** [open-power/ffs#7]
On a amd64 machine with 1GB of RAM, I get the following when running op-build:

```
TRACE: fpart --target /home/joel/op-build/output/host/usr/powerpc64le-buildroot-linux-gnu/sysroot/pnor/palmetto.pnor --partition-offset 33521664 --user 1 --name BACKUP_PART --value 2097152 --force
TRACE: fillPnorImage:: 0
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
fcp: errno : libffs.c(709) : (code=12) __ffs_buffer => Cannot allocate memory (errno=12)
```

**Comment by shenki** [comment]
I can run the tool standalone without issue. Could `buildpnor.pl` is consuming all of the memory?

**Comment by williamspatrick** [comment]
I don't think buildpnor.pl opens the files, so I don't see how it could be consuming all the memory.  Was op-build building anything in parallel?

@bradbishop - I thought I mentioned this issue before.  Anyone on your team able to propose a change to resolve this or at least terminate the build?

**Comment by williamspatrick** [comment]
Duplicate of #6.

**Comment by williamspatrick** [comment]
I don't see it so much of a problem that we require more memory than 1GB to build.  But, we should terminate the build when the PNOR isn't built correctly.  In #6 we had an issue where the resulting PNOR image was garbage as a result of this condition.

**Comment by ghost** [comment]
```
$ valgrind ./fcp/fcp ~/skiboot/skiboot.lid.xz /tmp/habanero.pnor:PAYLOAD --offset 0x00 --write --buffer 0x40000000
==30449== Memcheck, a memory error detector
==30449== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==30449== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==30449== Command: ./fcp/fcp /home/stewart/skiboot/skiboot.lid.xz /tmp/habanero.pnor:PAYLOAD --offset 0x00 --write --buffer 0x40000000
==30449== 
==30449== Warning: set address range perms: large range [0x5621040, 0x45621040) (undefined)
==30449== Warning: set address range perms: large range [0x59e41040, 0x99e41040) (undefined)
       0: PAYLOAD: write partition   100000/   52708
==30449== Warning: set address range perms: large range [0x59e41028, 0x99e41058) (noaccess)
==30449== Warning: set address range perms: large range [0x5621028, 0x45621058) (noaccess)
==30449== 
==30449== HEAP SUMMARY:
==30449==     in use at exit: 208 bytes in 9 blocks
==30449==   total heap usage: 25 allocs, 16 frees, 2,147,505,108 bytes allocated
==30449== 
==30449== LEAK SUMMARY:
==30449==    definitely lost: 24 bytes in 1 blocks
==30449==    indirectly lost: 0 bytes in 0 blocks
==30449==      possibly lost: 0 bytes in 0 blocks
==30449==    still reachable: 184 bytes in 8 blocks
==30449==         suppressed: 0 bytes in 0 blocks
==30449== Rerun with --leak-check=full to see details of leaked memory
==30449== 
==30449== For counts of detected and suppressed errors, rerun with: -v
==30449== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

aaand sure enough:

```
int __ffs_buffer(ffs_t * self, size_t size)
{
	assert(self != NULL);

        if (size == 0)
		size =  self->hdr->block_size;

        if (self->buf != NULL) {
		free(self->buf);
                self->buf_count = 0;
        }

        self->buf_count = size / self->hdr->block_size;
	size = self->buf_count * self->hdr->block_size;

        self->buf = malloc(size);
```

So, we're allocating 2GB of memory because that's what buildpnor.pl passed in here:
```
sub fillPnorImage {
....
            system("$g_fcpCmd $inputFile $i_pnorBinName:$eyeCatch --offset $offset --write --buffer 0x40000000");
```

it turns out we don't need this buffer at all, so I'm removing it.

Patch shortly.
